model/inference.py

```diff
@@ -0,0 +1,171 @@
+            user_df["friends_count"], 1
+        )
+        user_df["screen_name_length"] = user_df["screen_name"].str.len()
+        user_df["num_digits_in_screen_name"] = user_df["screen_name"].str.count(r"\d")
+        user_df["name_length"] = user_df["name"].str.len()
+        user_df["num_digits_in_name"] = user_df["name"].str.count(r"\d")
+        user_df["description_length"] = user_df["description"].str.len()
+        user_df.replace({False: 0, True: 1}, inplace=True)
+        user_df = user_df.select_dtypes("number").fillna(0.0)
+        return (user_df - user_mean) / user_std
+
+    def preprocessing_tweet(self, row):
+        rowlist = str(row).split()
+        rowlist = [word.strip() for word in rowlist]
+        rowlist = [
+            word if not word.strip().startswith("#") else "hashtagtag"
+            for word in rowlist
+        ]
+        rowlist = [
+            word if not word.strip().startswith("@") else "usertag" for word in rowlist
+        ]
+        rowlist = [word.lower() for word in rowlist]
+        rowlist = [re.sub(_URL_PATTERN, "urltag", word) for word in rowlist]
+        return " ".join(rowlist)
+
+    def vectorizing_tweet(self, tweets):
+        v = self.vectorizer.transform(tweets)
+        return v.A[np.newaxis, :]
+
+    def load_model(self, path, model_dict):
+        model = SOBOG(gpu=0, **model_dict)
+        model.load_state_dict(torch.load(path, map_location=torch.device("cpu")))
+        model.eval()
+        return model
+
+    def load_vectorizer(self, path):
+        with open(path, "rb") as fin:
+            vect = pickle.load(fin)
+        return vect
+
+    def generate_adj_matrix(self, tweet_df):
+        graph = nx.from_pandas_edgelist(tweet_df, "id", "parent_id")
+        graph.remove_node(0)
+        adj = nx.adjacency_matrix(graph, nodelist=tweet_df["id"].unique()).A
+        np.fill_diagonal(adj, 1.0)
+        return adj[np.newaxis, :]
+
+    def generate_user_post_matrix(self, tweet_df):
+        return (tweet_df["parent_id"] == 0).astype("int").values[np.newaxis, :]
+
+    def inference(self, user, tweet, adj, up):
+        user = torch.Tensor(user)
+        tweet = torch.Tensor(tweet)
+        adj = torch.Tensor(adj)
+        up = torch.Tensor(up)
+        user_pred, tweet_pred = self.model.forward(user, tweet, adj, up)
+        return user_pred
+
+    def predict(self, user_object, tweet_df):
+        user_df = self.create_user_dataframe(user_object)
+        user_df = self.preprocessing_user(user_df, self.user_mean, self.user_std)
+        user = user_df.values
+        adj = self.generate_adj_matrix(tweet_df)
+        tweet_df.drop_duplicates(subset="id", inplace=True)
+        tweets_text = tweet_df["text"].apply(self.preprocessing_tweet)
+        tweet = self.vectorizing_tweet(tweets_text)
+        up = self.generate_user_post_matrix(tweet_df)
+        user_pred = self.inference(user, tweet, adj, up)
+        return user_pred.item()
```

Comment:
Why `.item()`?

Reply:
Because `user_pred` is a `Tensor`

Reply:
Ohh yea the inference from original just return the tensor to print out to console, we need the value instead.